### DISCUSSION TITLE
Why do I need more than 1 connection when using nested transactions?

### DISCUSSION THREAD

**Discussion by kouwasi** [prisma/prisma#22711] [Q&A] [answered]
### Question

I'm using Prisma with PostgreSQL's RLS.

So I extended the PrismaClient like this to use RLS.

```ts
client.$extends({
  name: 'PostgresRLS',
  query: {
    async $allOperations({ args, query }) {
      try {
        const results = await client.$transaction([
          client.$executeRaw`SELECT set_config('request.jwt.claims', ${claims}, TRUE)`,
          query(args),
        ])

        return results[1]
      } catch (e) {
        console.error(e)
        throw e
      }
    },
  },
})
```

Then I connected to a database through pooler with `?connection_limit=1` option.
I expected to it works correctly, But PrismaClient thrown this error
```
Transaction API error: Transaction already closed: A commit cannot be executed on an expired transaction. The timeout for this transaction was 5000 ms, however 5107 ms passed since the start of the transaction. Consider increasing the interactive transaction timeout or doing less work in the transaction.
```

It happend only in using Transaction API when giving callback function.
```ts
prisma.$transaction(async (tx) => {
  tx.users.findMany({})   // raise error
})
```

How many connections I need? 

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: macOS
- Database: PostgreSQL 15
- Node.js version: v20.10.0

```
prisma                  : 5.8.1
@prisma/client          : 5.8.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.10.0
Query Engine (Node-API) : libquery-engine 78caf6feeaed953168c64e15a249c3e9a033ebe2 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 78caf6feeaed953168c64e15a249c3e9a033ebe2 (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.8.1-1.78caf6feeaed953168c64e15a249c3e9a033ebe2
Default Engines Hash    : 78caf6feeaed953168c64e15a249c3e9a033ebe2
Studio                  : 0.497.0
Preview Features        : multiSchema
```

**Comment by ludralph** [comment, marked as answer]
Hi @kouwasi 👋 

Thank you for raising this question.

The error message you're seeing suggests that the transaction is timing out. This could be due to the transaction taking longer than the configured timeout limit. You might want to consider increasing the timeout limit for transactions. See this section of the [documentation](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions:~:text=timeout%3A%20The%20maximum%20amount%20of%20time%20the%20interactive%20transaction%20can%20run%20before%20being%20canceled%20and%20rolled%20back.%20The%20default%20value%20is%205%20seconds.) on how to increase the timeout limit for your transaction.

Regarding the number of connections, it's important to note that the `?connection_limit=1` option you mentioned limits the connection pool to a single connection. This could potentially lead to issues if multiple transactions are trying to be executed concurrently, as they would be contending for the same connection. The appropriate number of connections would depend on your application's concurrency requirements and the workload you expect your database to handle.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
